feat: S3 presigned url 반환 api 구현

hyomee2 opened the pull request:
## 📌 관련 이슈
closed #23

## ✨ 작업 내용
이미지 업로드를 위한 s3 presigned url 반환 api를 구현했습니다.

## 📸 스크린샷(선택)
swagger와 postman를 통해, 생성된 presigned url과 이미지를 넣었을 때 s3 버킷에 이미지가 업로드 되는 것을 확인했습니다.
<img width="702" height="367" alt="image" src="https://github.com/user-attachments/assets/02752e74-a43e-4c3e-a3e5-7b5d10a3c8db" />

## 📚 레퍼런스 (또는 새로 알게 된 내용) 혹은 궁금한 사항들
[서버에서는 presigned url 발급만 담당하고, 실제 업로드는 클라이언트에서 해주어야 합니다!!]
클라이언트는 서버를 통해 Presigned URL을 발급받아 S3에 직접 업로드하고,
서버는 DB에 메타데이터 저장 + URL 발급만 담당합니다.

파일을 “보거나 다운로드”할 때도 동일하게 서버에 요청 → Presigned GET URL 발급 → 클라이언트가 S3에서 직접 가져오는 구조입니다.

hyomee2 committed:
refactor: 이미지 엔티티 리팩토링
- 불필요한 컬럼 삭제 및 컬럼 추가
- 디렉토리 변경

hyomee2 committed:
feat: S3 presigned url 생성 API 구현

hyomee2 committed:
Merge branch 'develop' of https://github.com/MuAring/backend into chore/#23-s3

ownue commented on src/main/java/com/example/muaring/config/S3Config.java:
```diff
@@ -0,0 +1,31 @@
+
+    private final S3Properties s3Properties;
+
+    @Bean
```
```@Bean(destroyMethod = "close")``` 로 빈 정리할 수 있도록 하는 건 어떨까요? (컨텍스트 종료 시 닫힘)

ownue commented on src/main/java/com/example/muaring/domain/file/service/ImageService.java:
```diff
@@ -0,0 +1,78 @@
+        PutObjectRequest putObjectRequest = PutObjectRequest.builder()
+                .bucket(s3Properties.s3().bucket())
+                .key(s3Key)
+                .build();
```
여기에 헤더 제약이 없는 걸 보완하면 좋을 것 같아요!

e.g.,
```java
PutObjectRequest put = PutObjectRequest.builder()
    .bucket(bucket)
    .key(s3Key)
    .contentType(expectedContentType)     // 예: image/jpeg, image/png
    .contentLength(fileSize)              // 실제 기대 파일 크기
    .contentMd5(base64Md5)                // 선택: 무결성
    .serverSideEncryption("aws:kms")      // 권장: KMS 사용 시
    .ssekmsKeyId(kmsKeyId)                // KMS 키
    .metadata(Map.of(
        "orig-filename", fileName,
        "image-type", imageType.name()
    ))
    .build();
```

ownue commented on src/main/java/com/example/muaring/domain/file/service/ImageService.java:
```diff
@@ -0,0 +1,78 @@
+        GetObjectRequest getObjectRequest = GetObjectRequest.builder()
+                .bucket(s3Properties.s3().bucket())
+                .key(image.getS3Key())
+                .build();
```
여기도 헤더 고정 어떤가용

e.g.,
```
GetObjectRequest get = GetObjectRequest.builder()
    .bucket(bucket)
    .key(image.getS3Key())
    .responseContentDisposition("inline; filename=\"" + safe(fileName) + "\"")
    .responseCacheControl("private, max-age=300")
    .build();
```

hyomee2 replied:
1) .responseContentDisposition("inline; filename=\"" + safe(fileName) + "\"")
- 저희는 파일을 다운로드 기능 제공은 안하고 앱에서 프로필 이미지를 표시하는 걸로 알고 있는데 그럴 경우 이건 필요없다고 하는데 어떻게 생각하시나요? 이미지를 아마 이미지 태그에서 직접 로드할 거 같은데 이 경우 브라우저는 content-disposition 헤더를 아예 무시해서 불필요할거 같아요!

2) responseCacheControl
- 저희는 지금 presigned url 을 생성하다 보니 어차피 10분 뒤에 만료돼서 캐시를 지정해도 의미가 없을 것 같은데 어떻게 생각하시나요? 혹시 필요하다면 배포 시에 cloudfront 이용해서 나중에 리팩토링하는 것도 좋을거 같아요!

ownue replied:
그럼 둘 다 빼도 좋을 것 같아요!!

ownue commented on src/main/java/com/example/muaring/domain/file/service/ImageService.java:
```diff
@@ -0,0 +1,78 @@
+
+        // 정의한 파일 정보에 서명하여 presigned URL 생성
+        PutObjectPresignRequest putObjectPresignRequest = PutObjectPresignRequest.builder()
+                .signatureDuration(Duration.ofMinutes(10))
```
만료 시간 설정값으로 분리하는 건 어떤지 제안 드립니당!!
e.g.,
```yaml
s3:
  presign:
    upload-exp-minutes: 10
    download-exp-minutes: 5
```
```java
Duration.ofMinutes(s3Properties.presign().uploadExpMinutes())
```

ownue commented on src/main/java/com/example/muaring/domain/file/controller/ImageController.java:
```diff
@@ -0,0 +1,39 @@
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
```
회원 기능 구현 후 권한 검증 필요 (```@PreAuthorize``` 또는 서비스단에서 처리)
- 업로드: 회원 본인 또는 그룹 ADMIN만
- 다운로드: 회원 전체

ownue commented on src/main/java/com/example/muaring/config/S3Properties.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@ConfigurationProperties(prefix = "cloud.aws")
+public record S3Properties (
```
파일명/크기/타입 검증 필요할 것 같습니당!!

hyomee2 replied:
서비스단에서 수행하도록 하겠습니다!

ownue commented on src/main/java/com/example/muaring/domain/file/service/ImageService.java:
```diff
@@ -0,0 +1,78 @@
+            case GROUP -> "group/";
+        };
+
+        String uniqueFileName = UUID.randomUUID() + "_"  + fileName;
```
보니까 여기에서 fileName(사용자 입력값)을 그대로 포함해서 저장하게 되는 것 같은데,
> 사용자 입력 fileName을 키에 그대로 포함시키면 XSS/로그 노출/경로 탐색 리스크
키는 prefix/uuid만 사용하고, 원본 파일명은 메타데이터로만 저장하세요.
혹은 안전한 슬러그 처리(safe(fileName)) 필수

찾아보니 이렇다고 해서 이 부분 수정해야 할 것 같습니당!!

ownue commented on src/main/java/com/example/muaring/domain/file/controller/ImageController.java:
```diff
@@ -0,0 +1,39 @@
+
+    private final ImageService imageService;
+
+    @GetMapping("/upload-url")
```
찾아보니 GET 메서드보다 POST + @Valid DTO 구현이 더 안전하다고 하는데 **POST + @Valid DTO(fileName, imageType, fileSize, contentType)** 이렇게도 고민해보면 좋겠습니다!!

ownue commented on src/main/java/com/example/muaring/domain/file/controller/ImageController.java:
```diff
@@ -0,0 +1,39 @@
+            @RequestParam Long fileSize) {
+        String presignedUploadUrl = imageService.generatePresignedUploadUrl(fileName, imageType, fileSize);
+        return ResponseEntity.ok(
+                ApiResponse.ok(presignedUploadUrl, "파일 업로드용 presigned URL이 성공적으로 생성되었습니다.")
```
지금 코드에서는 업로드 실패 시 presigned URL은 이미 발급되고, DB에는 Image 엔티티가 저장되기 때문에 데드 데이터 남을 수 있는 것 같아용!!! 업로드 완료 확인 후 저장되도록 수정하는 건 어떨까요!?

hyomee2 replied:
네 좋아요!! 그럼 우선은 presigned url 발급까지만 imageService에서 하고, 실제 메타데이터 저장은 프로필 수정 등에서 save 하도록 하겠습니다!

ownue reviewed (commented):
코드 리뷰 남겼습니다!

hyomee2 committed:
refactor: 컨텍스트 종료시 S3 리소스 정리

hyomee2 committed:
refactor: presigned url 만료시간 설정값으로 분리

hyomee2 committed:
refactor: presigned url 발급 로직 리팩토링
- S3 업로드 presigned url 발급 요청에 메타데이터 및 암호화 추가
- uniquefileName 명명 시 원본 파일명 사용 않도록 리팩토링
- 요청 시 GET -> POST 메서드로 수정

hyomee2 committed:
feat: 파일명/크기 검증 로직 추가
- 기존에 있던 파일타입 검증 로직을 service에서 validator로 분리했습니다.

hyomee2 committed:
chore: 충돌 해결

hyomee2 committed:
refactor: securityUtil 외부 인스턴스화 방지 추가

hyomee2 committed:
refactor: s3관련 설정값 분리 및 리팩토링

hyomee2 committed:
refactor: 이미지 업로드 presigned url 발급시 검증 로직 리팩토링
- 그룹 이미지의 경우 그룹 리더 검증 로직 추가
- dto에서 형식 검증, validator, service에서 비즈니스 로직 검증

hyomee2 committed:
chore: @Repository 어노테이션 추가

ownue reviewed (approved):
수고 많으셨어요!! 🥰🥰
